### PR TITLE
Add all foundation stash functions needed to support stashing in the app

### PR DESCRIPTION
**Pull request by iAmWillShepherd**
See tracking issue for more details - #6987

This PR brings together the functionality from the following PRs:
* Creating a stash entry - #7085
* Dropping a stash entry - #7123
* Applying a stash entry - #7036
* Finding the last stash entry created for a specific branch - #7089

These functions are needed to support the following workflows:

* Stashing changes on current branch
* Bringing changes over to new branch

### How do these functions relate to one another?

Both workflows start off from the same stash dialog.

**Stashing changes on current branch**
1. If there is an entry associated with the branch, drop it - `getLastDesktopStashEntryForBranch` -> `dropDesktopStashEntry`
1. Create stash entry for the **current** branch - `createDesktopStashEntry`
1. Perform the checkout
1. Check to see if new branch has an existing stash entry - `getLastDesktopStashEntryForBranch`

**Move to new branch and take changes there**
1. If there is an entry associated with the new branch, update the UI to inform user stash on new branch would be overwritten - `getLastDesktopStashEntryForBranch` -> `dropDesktopStashEntry`
1. Create stash entry for the **new** branch - `createDesktopStashEntry`
1. Perform the checkout
1. Apply the newly created stash immediately - `applyStashEntry`
1. Drop stash entry once applied - `dropDesktopStashEntry`

### CODE REVIEW

File: app/src/lib/git/stash.ts [hunk 1]
```diff
@@ -56,6 +71,7 @@ export async function getDesktopStashEntries(
 
     // if branch name is null, the stash entry isn't using our magic string
     if (branchName === null) {
+      // the stash entry isn't using our magic string, so skip it
```
Comment (shiftkey) [inline app/src/lib/git/stash.ts]: :art: can this comment be merged into the previous comment?

File: app/src/lib/git/stash.ts [hunk 2]
```diff
@@ -68,17 +84,82 @@ export async function getDesktopStashEntries(
...
+  stashSha: string
+) {
+  if (stashSha === '') {
+    // the drop sub-command behaves like pop when no stash is given
+    // so we return if given an empty string to avoid that
+    return
+  }
+
+  const entry = ['stash@{', stashSha, '}'].join('')
+  await git(['stash', 'drop', entry], repository.path, 'dropStashEntry')
```
Comment (shiftkey) [inline app/src/lib/git/stash.ts]: Are there any error cases here that you care about when dropping the stash? Or is this covered by future work?

File: app/src/lib/git/stash.ts [hunk 3]
```diff
@@ -68,17 +84,82 @@ export async function getDesktopStashEntries(
...
+ *
+ * To see the commit hash of stash entry, run
+ * `git log -g refs/stash --pretty="%nentry: %gd%nsubject: %gs%nhash: %H%n"`
+ * in a repo with some stash entries.
+ */
+export async function applyStashEntry(
+  repository: Repository,
+  stashSha: string
+): Promise<void> {
+  await git(
```
Comment (shiftkey) [inline app/src/lib/git/stash.ts]: Are there any error cases here that you care about when applying the stash? Or is this covered by future work?

File: app/src/lib/git/stash.ts (hunk 2 above)
Comment (iAmWillShepherd) [reply app/src/lib/git/stash.ts]: Added in df0bfe6

File: app/test/unit/git/stash-test.ts [hunk 4]
```diff
@@ -23,51 +28,209 @@ describe('git/stash', () => {
...
-  await FSE.appendFile(readme, '1')
-  await stash(repository, 'should get filtered')
-
-  await FSE.appendFile(readme, '2')
-  await stash(repository, 'should also get filtered')
-
-  // simulate stashing from Desktop
-  await FSE.appendFile(readme, '2')
-  await stash(repository)
+  await FSE.appendFile(readme, Math.random())
```
Comment (shiftkey) [inline app/test/unit/git/stash-test.ts]: There's a lint error here you might want to check:

```
  234:32  error  Math.random produces insecure random numbers. Use crypto.randomBytes() or window.crypto.getRandomValues() instead  insecure-random
```

File: app/src/lib/git/stash.ts (hunk 3 above)
Comment (iAmWillShepherd) [reply app/src/lib/git/stash.ts]: I added a base test case in ac67324 and I plan to handle other errors in a later PR. Many of the error cases I'd have to worry about will be taken care of in higher levels of the app by requiring certain invariants be met before calling these _risky_ git functions.

File: app/src/lib/git/stash.ts [hunk 5]
```diff
@@ -68,17 +88,89 @@ export async function getDesktopStashEntries(
...
+  const message = createDesktopStashMessage(branchName, tipSha)
+  await git(
+    ['stash', 'push', '-m', message],
+    repository.path,
+    'createStashEntry'
+  )
+}
+
+/**
+ * Removes the stash entry identified by `stashSha`
```
[the remark this answers is not in the document]
Comment (iAmWillShepherd) [reply app/src/lib/git/stash.ts]: Reverted this back to accepting the commit SHA of the stash entry and updated the docs to clarify this function only drops entries that exist.

**Comment by shiftkey** [comment]
I plan to merge this in tomorrow morning unless others have feedback to add

**Pull request merged**